torch/csrc/jit/tracer.h

```diff
@@ -245,24 +245,39 @@ void addInputs(Node *n, const char * name, std::array<bool, N> value) {
   throw std::runtime_error("Found an unsupported argument type in the JIT tracer. File a bug report.");
 }
 
-TORCH_API void postRecordTrace(Node* node, at::ArrayRef<Variable> outputs);
+namespace detail {
 
-inline void postRecordTrace(Node* node, at::ArrayRef<at::Tensor> tensors) {
-  postRecordTrace(node, fmap<Variable>(tensors));
-}
+void recordOutput(Node* node, const at::Tensor& tensor);
+void recordOutput(Node* node, const std::vector<at::Tensor>& list);
 
 template <
     typename T,
     typename = torch::enable_if_t<
-        (!std::is_convertible<torch::decay_t<T>, ArrayRef<Variable>>::value &&
-         !std::is_convertible<torch::decay_t<T>, ArrayRef<at::Tensor>>::value &&
-         !std::is_convertible<torch::decay_t<T>, Variable>::value)>>
-void postRecordTrace(Node* node, T&&) {
+        (!std::is_convertible<torch::decay_t<T>, at::TensorList>::value &&
+         !std::is_convertible<torch::decay_t<T>, at::Tensor>::value)>>
+void recordOutput(Node* node, T&&) {
   AT_ERROR(
       "Found an unsupported argument type ", at::demangle_type<T>(),
       " in the JIT tracer. File a bug report.");
 }
 
+
+} // namespace detail
+
+TORCH_API inline void postRecordTrace(Node *node) {};
+
+template<typename T, typename... Args>
+TORCH_API void postRecordTrace(Node* node, T&& out, Args&&... outputs) {
```

Comment:
When doing code generation, I've learned it is a bad idea to use two different technologies at once (e.g. both using the template system _and_ python code generation). We are already generating all calls to postRecordTrace. It would be much simpler to just write:

`[ "recordOutput(node, {});".format(r['name']) for r in declaration['returns'] ]`

In the code generator.

Reply:
Sure, I can do that.